specifications/unified-provider.org

```diff
@@ -28,57 +42,197 @@ Multipass) and potential future providers (Remote LXD, etc.).
+
+      Implementations should use the most performant approach available
+      for the given provider (i.e. bind mounting directory from container).
+
+      :param prime_dir: Directory to mount/sync prime directory to on host.
+
+      :raises SnapcraftProviderUnsupportedMethod: if provider does not support
+      syncing/mounting project's prime directory.
+      """
+
+    def snap(self, output_dir: pathlib.Path) -> List[pathlib.Path]:
+      """Snap project, building as required.
+
+      Write output snaps to host project directory.
+
+      :param output_dir: Directory to write snaps to.
+
+      :returns: Path to snap(s) created from build.
+      """
+
+      ...
+
+     def clean_parts(self, part_names: List[str]) -> None:
+      """Clean specified parts.
+
+      :param part_names: List of parts to clean.
+
+      :raises SnapcraftProviderUnsupportedMethod: if provider does not support
+      executing steps.
+      """
+
+      ...
+
+    def clean(self) -> None:
+      """Clean all artifacts of project and build environment.
+
+      Purges all artifacts from using the provider to build the
+      project.  This includes build-instances (containers/VMs) and
+      associated metadata and records.
+
+      This does not include any artifacts that have resulted from
+      a call to snap(), i.e. snap files or build logs.
+      """
+
+      ...
+
+    def shell(self, step: Optional[step.Step], part_name: Optional[str]) -> None:
+      """Launch an interactive shell to build-instance.
+
+      If available, load environment used for the given step and part_name.
+      Step and part_name must both be supplied for environment to be loaded,
+      otherwise both are ignored.
+
+      Note this method does not imply execute_step(), it is up to the caller
+      to launch the shell at the appropriate time.
+
+      :param step: step to load environment for.
+      :param part_name: part_name to load environment for.  If step is specified,
+      but not a part_name.
+
+      :raises SnapcraftProviderUnsupportedMethod: if provider does not support
+      providing a shell.
+      """
+
+      ...
+#+end_src
+
+*** Summary of changes against current provider:
+- =is_provider_ready()= replaces classmethod =ensure_provider()=.
+- =setup_provider()= becomes an instance method, supporting non-interactive
+  configuration, if possible.
+- Removes =mount_project()= interface. Provider will mount project as required.
 - Replace =pack_project()= with =snap()=, adding the implication that build steps
-  are to be executed as required (and packing may not be done by Snapcraft).
+  are to be executed as required, removing the implication that Snapcraft itself
+  invokes the packing (Snapcraft may just fetch the output of the provider).
 - Rename =clean()= to =clean_parts()=, to improve clarity of scope.
 - Rename =clean_project()= to =clean()=, to improve clarity of scope.
+- Context manager =__enter__()= and =__exit__()= are currently used to define
```

Comment:
This feel unbalanced, if all queries are related to the tinydb instance, why not query using tinydb directly (datastore instance) and then pass the relevant parameters into the provider?

Reply:
My original thought wass that the provider would know how to interpret the record(s) relevant to the provider.  But I can imagine another set of interfaces could manage that and pass it off as provider-specific kwargs.  That's probably preferable.

Reply:
I removed the datastore reference, I agree that shouldn't be there.

WRT 'imbalanced', are you referring to the shutdown of instance?  The other option was add a `close()` method and have the caller take care of it (probably with a `finally`).  I think I'd rather do that, but I mostly went this route because I couldn't come up with a good method name. :smirk:  Both approaches accomplish the same thing.  I'll add `close()` if you're good with it.

Reply:
Why not just "stop"? Opposite to "start"

Reply:
`start()` is not an interface..?

Reply:
Will come back with draft example